ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,851 @@
+      }
+      logSegment.setEndOffset(endOffset);
+    }
+    srcIndex.changeIndexSegments(indexSegmentFilesToAdd, indexSegmentsToRemove);
+  }
+
+  /**
+   * Gets a map of all the index segment files that cover the given {@code logSegmentNames}.
+   * @param logSegmentNames the names of log segments whose index segment files are required.
+   * @return a map keyed on log segment name and containing details about all the index segment files that refer to it.
+   */
+  private Map<String, NavigableMap<Offset, File>> getLogSegmentsToIndexSegmentsMap(List<String> logSegmentNames) {
+    Map<String, NavigableMap<Offset, File>> logSegmentsToIndexSegmentsMap = new HashMap<>();
+    for (String logSegmentName : logSegmentNames) {
+      logSegmentsToIndexSegmentsMap.put(logSegmentName, getIndexSegmentDetails(logSegmentName));
+    }
+    return logSegmentsToIndexSegmentsMap;
+  }
+
+  // cleanup() helpers
+
+  /**
+   * Waits until the ref counts of the all the log segments that need to be cleaned up reach 0.
+   * @throws InterruptedException if the wait is interrupted.
+   */
+  private void waitTillRefCountZero() throws InterruptedException {
+    if (isActive) {
+      long singleWaitMs = Math.min(500, WAIT_TIME_FOR_CLEANUP_MS);
+      long waitSoFar = 0;
+      List<String> segmentsUnderCompaction = compactionLog.getCompactionDetails().getLogSegmentsUnderCompaction();
+      for (String segmentName : segmentsUnderCompaction) {
+        while (srcLog.getSegment(segmentName).refCount() > 0 && waitSoFar < WAIT_TIME_FOR_CLEANUP_MS) {
+          time.sleep(singleWaitMs);
+          waitSoFar += singleWaitMs;
+        }
+      }
+    }
+  }
+
+  /**
+   * Cleans up all the log and index segments. This involves:
+   * 1. Dropping the log segments that are no longer relevant.
+   * 2. Deleting all associated index segment and bloom files.
+   * @param recovering {@code true} if this function was called in the context of recovery. {@code false} otherwise.
+   * @throws IOException
+   */
+  private void cleanupLogAndIndexSegments(boolean recovering) throws IOException {
+    CompactionDetails details = compactionLog.getCompactionDetails();
+    List<String> segmentsUnderCompaction = details.getLogSegmentsUnderCompaction();
+    for (int i = 0; i < segmentsUnderCompaction.size(); i++) {
+      String segmentName = segmentsUnderCompaction.get(i);
+      if (srcLog.getSegment(segmentName) != null) {
+        PersistentIndex.cleanupIndexSegmentFilesForLogSegment(dataDir.getAbsolutePath(), segmentName);
+        srcLog.dropSegment(segmentName, recovering || i >= numSwapsUsed);
+      }
+    }
+  }
+
+  /**
+   * Resets all the internal data structures and wipes the slate clean for the next compaction.
+   */
+  private void resetStructures() {
+    tgtLog = null;
+    tgtIndex = null;
+    numSwapsUsed = 0;
+    recoveryStartToken = null;
+  }
+
+  /**
+   * Ends compaction by closing the compaction log if it exists.
+   */
+  private void endCompaction() {
```

Comment:
It closes the compaction log and sets the internal reference of the compactor to null.